### PR TITLE
Remove SdkContext from core test framework, and do some enhancement.

### PR DESCRIPTION
**Pull request by sima-zhu**
1. The track two library azconfig is already depending on test-common.
2. Other track two library will move to test-common using the test framework. #3208 
3. Changes of the PR:
a. Make SdkContext final as helper class, and make the common methods in SdkContext as static.
b. Refactor Azconfig to use the static method in SdkContext.
c. Make TestResourceNamer to public class, so TestBase has access to it.

### CODE REVIEW

**Comment by conniey** [comment]
Why the change to make this a static class/static methods? I had moved this to a nonstatic class with TestResourceNamer inside because its only function was to get random names in SdkContext.

**Comment by sima-zhu** [comment]
> Why the change to make this a static class/static methods? I had moved this to a nonstatic class with TestResourceNamer inside because its only function was to get random names from SdkContext. You're using it in the same way here, except with static.

@conniey The java doc for this class indicates it is a util class for common methods. I have several other methods placed in the class, so it is not only for generating the random name. It might not necessary to have the testmode and testname for other methods. That's why it made it static and remove the constructor.

**Comment by conniey** [comment]
@sima-zhu You can make methods inside the class static without making _all_ of the methods static. So you can have `public static String replaceHost() { }` and `public String generateRandomName()` in the same class.

**Comment by sima-zhu** [comment]
> @sima-zhu You can make methods inside the class static without making all of the methods static. So you can have public static String replaceHost() { } and public String generateRandomName() in the same class.
@conniey 
I know it is easy to make both work in SdkContext. I prefer to make it static class as static saves memory, no need to instantiate it, which make more senses to me. Do I miss any context which is necessary to make the SdkContext non-static?

**Comment by conniey** [comment]
I made SdkContext non static from the original code because having a static method with static TestNamer for getting random test names limits the number of instances we can run in parallel. The SdkContext.randomNames() can probably be removed and have those methods added to TestResourceNamer, which would be responsible for naming things.
https://github.com/Azure/azure-sdk-for-java/pull/3472/files#diff-796d9582e6fd2ee76b457d65f6a1d7a5R35

File: common/azure-common-test/src/main/java/com/azure/common/test/utils/TestConstant.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+// Licensed under the MIT License.
+
+package com.azure.common.test.utils;
+
+/**
+ * Constant for test classes.
+ */
+public final class TestConstant {
+    public static final int DEFAULT_BUFFER_LENGTH = 1024;
+}
```
Comment (JonathanGiles) [inline common/azure-common-test/src/main/java/com/azure/common/test/utils/TestConstant.java]: I see no need for this class - the value could be a private static field inside SdkContext

File: common/azure-common-test/src/main/java/com/azure/common/test/utils/SdkContext.java [hunk 2]
```diff
@@ -3,58 +3,151 @@
...
-     * </ul>
-     *
-     * @param testMode The test for this context.
-     * @param recordedData The data to persist or read any variables names to or from.
-     */
-    public SdkContext(TestMode testMode, RecordedData recordedData) {
-        Objects.requireNonNull(recordedData);
-        nameGenerator =  new TestResourceNamer("", testMode, recordedData);
-    }
+public final class SdkContext {
```
Comment (JonathanGiles) [inline common/azure-common-test/src/main/java/com/azure/common/test/utils/SdkContext.java]: I don't know why this class is called SdkContext anymore. It seems like it is now just a TestUtils class or similar.

File: common/azure-common-test/src/main/java/com/azure/common/test/utils/SdkContext.java [hunk 3]
```diff
@@ -3,58 +3,151 @@
...
+     * @param namer the test resource random namer which is related to recorded data
      * @param prefix the prefix to be used if possible
      * @param maxLen the maximum length for the random generated name
      * @return the random name
      */
-    public String randomResourceName(String prefix, int maxLen) {
-        return nameGenerator.randomName(prefix, maxLen);
+    public static String randomTestResourceName(TestResourceNamer namer, String prefix, int maxLen) {
+        return namer.randomName(prefix, maxLen);
     }
```
Comment (JonathanGiles) [inline common/azure-common-test/src/main/java/com/azure/common/test/utils/SdkContext.java]: At this point it is more verbose for people to call this method that to simply call the code that this method calls. I would suggest removing this method entirely.

File: common/azure-common-test/src/main/java/com/azure/common/test/http/RecordClient.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (conniey) [inline common/azure-common-test/src/main/java/com/azure/common/test/http/RecordClient.java]: This class exists already: $\common\azure-common-test\src\main\java\com\azure\common\test\policy\RecordNetworkCallPolicy.java

File: common/azure-common-test/src/main/java/com/azure/common/test/models/RecordedData.java [hunk 5]
```diff
@@ -88,4 +88,13 @@ public String removeVariable() {
             return variables.remove();
         }
     }
+
+    /**
+     * Getter method for {@link RecordedData#networkCallRecords}
+     *
+     * @return network call records in session record.
```
Comment (conniey) [inline common/azure-common-test/src/main/java/com/azure/common/test/models/RecordedData.java]: I didn't add this method because its use is covered by the existing methods above such as findFirstAndRemoveNetworkCall and addRecord. Is there any reason to expose the underlying LinkedList?

File: common/azure-common-test/src/main/java/com/azure/common/test/utils/SdkContext.java [hunk 6]
```diff
@@ -3,58 +3,151 @@
...
+
+    /**
+     * Replace the text to the one defined in textReplacementRules.
+     *
+     * @param text the text which will apply the text replacement rules
+     * @param textReplacementRules the rule of text replacement
+     * @return the updated text
+     */
+    public static String applyReplacementRule(String text, Map<String, String> textReplacementRules) {
```
Comment (conniey) [inline common/azure-common-test/src/main/java/com/azure/common/test/utils/SdkContext.java]: Why was this moved from the PlaybackClient? I had moved the logic in there because its only usecase, from what I saw in the tests were to replace contents in a recorded HttpResponse. I don't see how this is a generic utility method.

File: common/azure-common-test/src/main/java/com/azure/common/test/utils/SdkContext.java [hunk 7]
```diff
@@ -3,58 +3,151 @@
...
+    }
+
+    /**
+     * Async API for extract and rebuild response data.
+     *
+     * @param response the response to extract and rebuild
+     * @param textReplacementRule rules of rebuilding the response data
+     * @return the response body map of name and content
+     */
+    public static Mono<Map<String, String>> extractResponseData(final HttpResponse response, final Map<String, String> textReplacementRule) {
```
Comment (conniey) [inline common/azure-common-test/src/main/java/com/azure/common/test/utils/SdkContext.java]: Similarly, this already exists in RecordNetworkCallPolicy.

**Comment by conniey** [comment]
It might be worth exploring why a method or class exists for our existing test infrastructure and what already exists in azure-common-test. When I looked at our existing test infrastructure, parts of it seemed disjoint, and put together for ease-of-use rather than good design.

**Comment by sima-zhu** [comment]
Discussed with Connie, we will put the methods into its most related class. Use TestRourceNamer direcly instead of calling SdkContext. Will review all methods in SdkContext again.

File: common/azure-common-test/src/main/java/com/azure/common/test/utils/SdkContext.java (hunk 6 above)
Comment (sima-zhu) [reply common/azure-common-test/src/main/java/com/azure/common/test/utils/SdkContext.java]: I saw this one appear at different places in old interceptor, so I decided to move out. I double check with RecordNetworkCallPolicy, and it does not use the same method over there. Therefore, there is no need to maek it common.  Why the RecordNetworkCallPolicy does not have the same use?

File: common/azure-common-test/src/main/java/com/azure/common/test/utils/TestConstant.java (hunk 1 above)
Comment (sima-zhu) [reply common/azure-common-test/src/main/java/com/azure/common/test/utils/TestConstant.java]: The new commits will remove the class as the Constant is only used at one place.

File: common/azure-common-test/src/main/java/com/azure/common/test/http/RecordClient.java (hunk 4 above)
Comment (sima-zhu) [reply common/azure-common-test/src/main/java/com/azure/common/test/http/RecordClient.java]: One question here, why it has two different way of handling the recorded network, (e.g. PlaybackClient in playback mode and RecordNetworkCallPolicy in recoreded mode)?

File: common/azure-common-test/src/main/java/com/azure/common/test/utils/SdkContext.java (hunk 6 above)
Comment (conniey) [reply common/azure-common-test/src/main/java/com/azure/common/test/utils/SdkContext.java]: The method signature may not be the same, but the functionality is. When you create a PlaybackClient, you pass in a set of textReplacementRules upfront, so the client applies the transformation to each response that is played back.

File: common/azure-common-test/src/main/java/com/azure/common/test/models/RecordedData.java (hunk 5 above)
Comment (sima-zhu) [reply common/azure-common-test/src/main/java/com/azure/common/test/models/RecordedData.java]: Deleted the one.

Comment (conniey) [reply common/azure-common-test/src/main/java/com/azure/common/test/models/RecordedData.java]: I saw its usage in the old test cases and saw it as breaking the encapsulation OO design principle by unnecessarily exposing the underlying data type for ease of use.
https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)

File: common/azure-common-test/src/main/java/com/azure/common/test/http/RecordClient.java (hunk 4 above)
Comment (conniey) [reply common/azure-common-test/src/main/java/com/azure/common/test/http/RecordClient.java]: PlaybackClient is when we don't have a live connection to an Azure Service. So we read a session-record from the test/resources/session-records/*.json location. And for each network call, play back a previously copied response.
RecordNetworkCallPolicy is for when we have a connection to an Azure Service. We save each HTTP request and response so that we can verify its results or play it back later.

Comment (sima-zhu) [reply common/azure-common-test/src/main/java/com/azure/common/test/http/RecordClient.java]: Thanks for the explanation. Really helpful.

File: common/azure-common-test/src/main/java/com/azure/common/test/models/RecordedData.java (hunk 5 above)
Comment (sima-zhu) [reply common/azure-common-test/src/main/java/com/azure/common/test/models/RecordedData.java]: Good point. Removed it.
